### PR TITLE
[IMP] core: ease testing upgrade scripts in custom modules

### PR DESCRIPTION
**Pull request by erl-odoo**
Upgrade (aka migration) scripts are a core part of Odoo, allowing
database manipulations for modules during version changes.

Any module, including custom ones can run upgrade scripts, even if the
`--upgrade-path` flag (and with it, the `odoo.upgrade` sub-module) is
not present. Currently only the "standard" modules benefit of easy
upgrade script testing. Any custom modules that want to run tests of
their upgrades have to import the tests in the usual `tests` folder,
which is not ideal.

Therefore, to allow TDD and programmatic testing of upgrade scripts in
custom modules, the test discovery is here modified to also parse the
module's `migrations` and `upgrades` sub-modules for tests.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by erl-odoo** [comment]
Follow up on #122421

File: odoo/modules/module.py [hunk 1]
```diff
@@ -501,16 +495,36 @@ def _get_tests_modules(path, module):
...
+    # if not in `odoo.upgrade`, try for upgrade submodules, which could be used by custom modules
+    upgrade_submodules = [f"odoo.addons.{module}.{submodule}" for submodule in ("migrations", "upgrades")]
+    for module_name in upgrade_submodules:
+        try:
+            mod = importlib.import_module(module_name)
+            test_modules += list(yield_test_modules(mod, "*/tests/test_*.py"))
+        except ImportError:
+            pass
+
+    return test_modules
```
[the remark this answers is not in the document]
Comment (aj-fuentes) [reply odoo/modules/module.py:527]: I think the version without nested function is better.

**Comment by erl-odoo** [comment]
@KangOl any news when we can finish this?

**Comment by KangOl** [comment]
@robodoo r+

**Comment by pedrobaeza** [comment]
Hi, since this commit, we are getting a weird recursion limit with the OCA module `base_maintenance`:

https://github.com/OCA/maintenance/pull/359
https://github.com/OCA/maintenance/actions/runs/6271592917/job/17031481881?pr=359#step:8:2581

Can you please check why? I suspect it gets a namespace colliding due to your custom `odoo.addons.base.maintenance` namespace used for migration.

**Comment by KangOl** [comment]
It is really weird. `odoo.addons.base.maintenance` != `odoo.addons.base_maintenance`.

And the traceback recursion loop is directly inside cpython code.
I will investigate about it.

**Pull request closed without merging**